### PR TITLE
Support hexagon tiles in `GridMap`

### PR DESCRIPTION
**Pull request by ghost**
This adds support for hexagonal grids in the `GridMap` node with the same offset and layout semantics as `TileMap`.

Closes https://github.com/godotengine/godot-proposals/issues/4337

### CODE REVIEW

**Review by fire-forge — COMMENTED** [review]
I tested it and it works really well. Just a few things need to be changed/fixed:
* The editor grid drawing should be changed to respect `cell_shape` and `cell_offset_axis` settings like the TileMap editor grid does.
* `cell_offset_axis` only works in hexagon mode, so you can't use offset squares like you can in TileMaps.
* The placement position in the editor is off by about half of a cell. This is more apparent when you go into the Top Orthogonal camera mode when drawing on the GridMap. Disabling `cell_center_x/y/z` makes it off by half of a cell in the opposite direction.

**Comment by octetdev2** [comment]
> The editor grid drawing should be changed to respect cell_shape and cell_offset_axis settings like the TileMap editor grid does.

Was going to split this out into a follow up pull request, but squashed it into here.

> cell_offset_axis only works in hexagon mode, so you can't use offset squares like you can in TileMaps.

Yes, there is no half offset square shape yet.
But the inspector should disable the options when not applicable.

> The placement position in the editor is off by about half of a cell. This is more apparent when you go into the Top Orthogonal camera mode when drawing on the GridMap. Disabling cell_center_x/y/z makes it off by half of a cell in the opposite direction.

I can't reproduce this? The transform for cursor instance and rendering instance should be identical.

**Comment by fire-forge** [comment]
> I can't reproduce this? The transform for cursor instance and rendering instance should be identical.

I meant that the mouse cursor position vs. which cell it decides you're hovering over is about half a cell off, not the cursor and rendering instance being off. I'll try to get a video of it, and I'll check if it happened before this PR too.

**Comment by octetdev2** [comment]
>Nevermind, glad to hear it's fixed.

Actually, I lost the grid offset changes when I rebased and squashed about an hour ago... never the less, easy fix.

File: modules/gridmap/grid_map.cpp [hunk 1]
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
...
+		if (cell_offset_axis == GridMap::CELL_OFFSET_AXIS_HORIZONTAL) {
+			ret.z *= 0.75;
+		} else { // CELL_OFFSET_AXIS_VERTICAL
+			ret.x *= 0.75;
+		}
+	}
+
+	// Add the offset
+	Vector3 offset = Vector3(
+			cell_size.x * 0.5 * int(center_x),
```
Comment (octetdev2) [inline modules/gridmap/grid_map.cpp:501]:
```suggestion
			0.5 * int(center_x),
```

File: modules/gridmap/grid_map.cpp [hunk 2]
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
...
+			cell_size.y * 0.5 * int(center_y),
```
Comment (octetdev2) [inline modules/gridmap/grid_map.cpp:502]:
```suggestion
			0.5 * int(center_y),
```

File: modules/gridmap/grid_map.cpp [hunk 3]
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
...
+			cell_size.z * 0.5 * int(center_z));
```
Comment (octetdev2) [inline modules/gridmap/grid_map.cpp:503]:
```suggestion
			0.5 * int(center_z));
```

**Review by ghost — COMMENTED** [review]
Oops, multiplying this twice would be the cause of that extra offset (which was only observable depending on cell size).

**Comment by octetdev2** [comment]
- Figured out the property validator and disabled the properties for square cells.
- Also squashed the off by one error when picking (off by one extra pass of multiplication 😆).

**Comment by Wasabi-Cheetah** [comment]
I'm just going to post this here for anyone else looking for a hexagon grid system for a sphere as well. https://stackoverflow.com/questions/46777626/mathematically-producing-sphere-shaped-hexagonal-grid

**Comment by soft-circles** [comment]
I would love to see this feature merged, any chance this will make it in during the Godot 4 beta?

**Comment by Calinou** [comment]
> I would love to see this feature merged, any chance this will make it in during the Godot 4 beta?

[Godot 4.0 is in feature freeze, so any features of a significant size are postponed to future 4.x releases.](https://godotengine.org/article/release-management-4-0-and-beyond)

**Comment by gulagkulak** [comment]
@octetdev2 seems to have disappared from the internet (I hope he's ok) so, uh, can someone please take over development of this feature?

**Comment by AThousandShips** [comment]
Salvaged in: #85890

**Pull request closed without merging**